### PR TITLE
Add python-wasi on run-code

### PR DESCRIPTION
**Pull request by WangGithub0**
Add python-wasi in run-code, and also addressing typeScript type declarations for it.
#312 

image

image

### CODE REVIEW

File: src/components/CodeHeader.tsx [hunk 1]
```diff
@@ -16,7 +16,7 @@ import { TbCopy, TbDownload, TbRun, TbExternalLink } from "react-icons/tb";
 
 import { download, formatAsCodeBlock } from "../lib/utils";
 import { useAlert } from "../hooks/use-alert";
-import { isRunnable, runCode } from "../lib/run-code";
+import { isRunnable, isRunnableOnServer, runCode } from "../lib/run-code";
```
Comment (humphd) [inline src/components/CodeHeader.tsx:19]: I wonder if we should rename `isRunnable` to `isRunnableInBrowser` to match what you're doing here.

File: src/lib/run-code.ts [hunk 2]
```diff
@@ -11,7 +11,9 @@ globalThis.__esbuildWasmLoaded = false;
 
 const supportedJS = ["js", "javascript"];
 const supportedTS = ["ts", "typescript"];
-const supportedLanguages = [...supportedJS, ...supportedTS];
+const supportedPY = ["py", "python"];
+const supportedLanguages = [...supportedJS, ...supportedTS, ...supportedPY];
```
Comment (humphd) [inline src/lib/run-code.ts:15]: `supportedBrowserLanguages` maybe?

File: src/lib/run-code.ts [hunk 3]
```diff
@@ -105,6 +115,41 @@ async function runJavaScript(code: string) {
   }
 }
 
+async function runPython(code: string) {
+  const { WASI } = await import("@antonz/runno");
+  const url = "https://unpkg.com/@antonz/python-wasi/dist/python.wasm";
```
Comment (humphd) [inline src/lib/run-code.ts:120]: Can we get this from node_modules vs. unpkg.com?  Vite allows you to get a URL for a WASM https://vitejs.dev/guide/features#accessing-the-webassembly-module

**Review by humphd — CHANGES_REQUESTED** [review]
This is amazing, such great work.

**Comment by tarasglek** [comment]
thats exciting, lets make the run button run it instead of showing menu when no server option

**Comment by tarasglek** [comment]
love how small this change is

File: src/lib/run-code.ts (hunk 3 above)
Comment (WangGithub0) [reply src/lib/run-code.ts]: I change to use node_modules and can build successfully on my local, but fail in [CloudFlare](https://dash.cloudflare.com/?to=/:account/pages/view/chatcraft-org/2ae02daa-e807-47a9-80e1-858c95ad4f84), could you help me check the log on CloudFlare

Comment (humphd) [reply src/lib/run-code.ts]:
```

22:37:22.763 | Deploying your site to Cloudflare's global network...
-- | --
22:37:26.503 |  
22:37:26.618 | ✘ [ERROR] Error: Pages only supports files up to 26.2 MB in size
22:37:26.619 |  
22:37:26.619 | assets/python-vjebDg3S.wasm is 26.3 MB in size

```

So we're going to have to load it from a CDN.  I guess go back to using unpkg.com.

**Review by humphd — APPROVED** [review]

**Review by kliu57 — APPROVED** [review]
I tested it and it works on running python code from both user's own python code and chatgpt's reply python code. Only thing I found was it will output an error if you try to run any python code that is asking for user input (because obviously we have no terminal to take in user's input I suppose)

Maybe something for future follow up PRs could be to print more details error messages when such fails occur. Right now it just says something like this

```
Error: Script execution failed

ySe/</</<@https://60283562.console-overthinker-dev.pages.dev/assets/index-1LG397X1.js:579:1095
```

**Pull request merged**